packages/actor-rdf-resolve-hypermedia-links-queue-wrapper-info-occupancy/lib/LinkQueueLogger.ts

```diff
@@ -0,0 +1,209 @@
+import type { ILinkQueue, ILink } from '@comunica/bus-rdf-resolve-hypermedia-links-queue';
+import { LinkQueueWrapper } from '@comunica/bus-rdf-resolve-hypermedia-links-queue';
+import type { Logger } from '@comunica/types';
+import { PRODUCED_BY_ACTOR } from '@comunica/types-link-traversal';
+import type { Algebra } from 'sparqlalgebrajs';
+
+/**
+ * A link queue that log information about event happening in the queue
+ */
+export class LinkQueueLogger extends LinkQueueWrapper {
+  public readonly query: Algebra.Operation;
+  private readonly logger: Logger;
+  private readonly reachabilityRatio: IReachabilityRatio = {
+    pushEvent: {},
+    popEvent: {},
+  };
+
+  public static readonly LINK_QUEUE_EVENT_NAME = '<Link queue occupancy>';
+  public static readonly LINK_QUEUE_DIDNT_START_EMPTY_MESSAGE = 'the link queue didn\'t start empty';
+
+  /**
+   *
+   * @param {ILinkQueue & IOptionalLinkQueueParameters} linkQueue - The link queue with optional parameters
+   * @param {Algebra.Operation} query - The current query
+   * @param {Logger} logger - The logger where the events are output
+   */
+  public constructor(linkQueue: ILinkQueue & IOptionalLinkQueueParameters, query: Algebra.Operation, logger: Logger) {
+    super(linkQueue);
+    this.logger = logger;
+
+    if (!linkQueue.isEmpty()) {
+      this.logger.warn(
+        LinkQueueLogger.LINK_QUEUE_EVENT_NAME,
+        {
+          message: LinkQueueLogger.LINK_QUEUE_DIDNT_START_EMPTY_MESSAGE,
+        },
+      );
+    }
+    this.query = JSON.parse(JSON.stringify(query));
+
+    Object.freeze(this.query);
+  }
+
+  /**
+   * Helper function to get the reachability criteria of a link
+   * @param {ILink} link - Current link
+   * @returns {[string, object | undefined] | null}
+   * The reachability criteria with extra information about it if available
+   */
+  private static getReachability(link: ILink): [string, object | undefined] | null {
```

Comment:
This method return type is very confusing. Can you simplify it? What do string/object/undefined/null mean?
We prefer `undefined` over `null`. Using them both is very confusing.

Reply:
I use null because when converted in a JSON it still preserve the field. But I will revisit the interface and this function.